lz4 for rosbag

bcharrow opened the pull request:
I'd really like rosbag to have a fast compression option.  The use case is for larger bag files that take forever to process when you use bz2; while bz2 is great for long term storage, it can be unreasonably slow when you want to repeatedly process a bag.  lz4 (https://code.google.com/p/lz4/) seems like a pretty good option for this use case, so I made a prototype in roscpp.  The results are encouraging, though my implementation isn't complete so the changes shouldn't be merged just yet.

To measure performance, I used this [big bagfile](http://infinity.csail.mit.edu/data/2011/2011-01-18-06-37-58.bag).  Info about it is [here](http://infinity.csail.mit.edu/data/2011/utilities/2011-01-18-06-37-58/2011-01-18-06-37-58.bag.info).  See http://projects.csail.mit.edu/stata/downloads.php for details.

My machine: i7 sandy bridge, 16GB DDR3 RAM, intel SSD, compiled everything scratch using Release build type.

Create compressed bag file

| Type | Real (sec) | User (sec) | Size (%) |
| --- | --: | --: | --: |
| bz2 (python) | 2986.8 | 2954.2 | 17.16 |
| bz2 (C++) | 3050.0 | 3009.4 | 17.16 |
| lz4 (C++) | 51.2 | 21.6 | 25.83 |

Create decompressed bag file from compressed one

| Type | Real (sec) | User (sec) |
| --- | --: | --: |
| bz2 (python) | 333.1 | 303.0 |
| bz2 (C++) | 351.9 | 307.3 |
| lz4 (C++) | 88.8 | 7.5 |

Real / user are from time command.  bz2 (python) is current "rosbag compress" where as bz2 (c++) is the new executable I added.  These results are pretty consistent with what the lz4 page has.

A couple of notes:
1. As far as I can tell, lz4 isn't available as a system dependency, so I hackishly added the files to rosbag_storage.
2. I added C++ executables for compress/decompress so I wouldn't have to add things in python, but given the difference between c++/python for bz2, it looks like compress/decompress for lz4 should also be done in python.
3. I used the [lz4 streaming format](https://docs.google.com/document/d/1gZbUoLw5hRzJ5Q71oPRN6TO4cRMTZur60qip-TE7BhQ) to encode multiple chunks, although no checksums are currently present.
4. I forked form 1.9.54

If there's a chance this can eventually get merged, I'll finish the implementation, add support for python and update the rosbag command.

Thoughts?

forouher commented:
+1

Rosbagging topics like pointclouds was always a hassle for me. Uncompressed rosbag maxed out my non-SSD hard drive, whereas bz2 compression didn't go well with my CPU (which had other things to do ;)

So this looks like a promising middle ground.

po1 commented:
What is the motivation for `lz4` in particular?
It is my understanding that it performs really fast, sacrificing compression ratio.

`bzip2` is notoriously slow for both compression _and_ decompression. Compression usually only happens once per bagfile, but decompression will take place whenever the bag file is opened, and this is usually when we need speed.

Are there any benchmarks with other algorithms, such as `zlib`? The latter has the advantage of being quite ubiquitous, and does not require duplicating its source in the ROS source tree.

Aside for the specific algorithm used, +1 for using anything else than `bzip2`

bcharrow commented:
I picked lz4 because it's substantially faster for both compression and decompression than the alternatives I found, zlib included.  https://code.google.com/p/lz4/ has the comparison you asked for; lz4's decompression speed is about 6 times faster than zlib.  I'm inclined to suggest that this difference in speed merits the duplication of lz4's source, but I can see both sides.

Although it's still duplication, would it be better to create a separate pure CMake project (a la console_bridge) for compressing / decompressing lz4 streams?  rosbag_storage would then depend on that.

Google's [snappy](https://code.google.com/p/snappy/) or [LZO](http://www.oberhumer.com/opensource/lzo/) are alternatives that are available and faster than zlib.  However, lz4 still seems to faster than both with no real difference in compression ratio.

dirk-thomas commented:
Is the (now available) Ubuntu package (http://packages.ubuntu.com/source/trusty/lz4 / http://packages.ubuntu.com/source/saucy/lz4) new enough to be used rather then the embedded version? If yes, can you please update the PR to use that package (including a rosdep rule for it)?

dirk-thomas commented:
Regarding the added executables `compress` and `decompress`: how are they different to the existing commands `rosbag compress` / `rosbag decompress`? Wouldn't it make sense to extend the existing commands with options to choose the compression type instead of adding new executables?

dirk-thomas commented:
Could you please also address the `todos` mentioned in the new code and implement the additional checks?

bcharrow commented:
It looks like the new ubuntu package is recent enough, I'll update the PR accordingly.  This doesn't address python bindings, though.  There's a [pip package](https://pypi.python.org/pypi/lz4), but it doesn't use the system version of lz4.  How this sort of thing is handled elsewhere?

I added the compress / decompress executables to see if they'd be faster than the compress / decompress options in python.  It looks like there's no performance improvement, so I agree it makes more sense to extend the existing rosbag commands to choose the compression type.

In my next update, I'll also address the TODOs, and add support for python.

dirk-thomas commented:
Great, thank you.

I am not sure if there is a Python implementation around - or a ready to use binding. But we should aim to keep the Python version of rosbag feature equal to the C++ version.

dirk-thomas commented:
@bcharrow It would be great if you could address the pending details so that it can be released into _Indigo_.

ros-pull-request-builder commented:
Can one of the admins verify this patch?

bcharrow commented:
@dirk-thomas I plan on getting everything finished shortly (i.e., a week or two).  Will this work for indigo?

dirk-thomas commented:
That sounds good. It can definitely go into _indigo-devel_. I just want to make sure it gets in early enough to have some soak time before the release. Thanks for working on it.

bcharrow commented:
I've now implemented C++/python bindings for the [lz4 streaming format](https://docs.google.com/a/seas.upenn.edu/document/d/1gZbUoLw5hRzJ5Q71oPRN6TO4cRMTZur60qip-TE7BhQ/edit), incorporated them into C++/python rosbag, and added options to the relevant command line tools.  To make the commit history a little cleaner, I squashed, and rebased off of indigo-devel.  This set of changes uses the system lz4 library for block compression.

The new thing in these commits is the roslz4 package which contains python/C++ APIs for lz4 streaming similar to those for bzip2.  This includes breaking up input into chunks, computing checksums to validate data, and a stream descriptor header.  I made the package because I figured having the headers/checksums was a good idea for long term data storage in bags, and as far as I can tell there isn't an open source library which implements the streaming format (though it is a requested feature https://code.google.com/p/lz4/issues/detail?id=42).  I had to add the source for the xxhash library because it's used to compute the checksums, and not available as a system dependence.  roslz4 does follow the lz4 stream spec, so if at some point a system library becomes available, the package could be removed.

Using roslz4, the changes to rosbag are much simpler; I don't think there's anything too complicated going on there.

Although I didn't commit unit tests, I have done some testing.  One example: https://gist.github.com/bcharrow/11016627 is a file compressor/decompressor implemented with roslz4.  I've successfully compressed/decompressed files produced by the lz4c tool (apt-get install libz4-tool).  This gives some confidence that the standard is being followed.

Two things I still need to do:
- add the read streaming functions for LZ4Stream (i.e., startRead(), read(), stopRead())
- Add a rosdep rulre for lz4; for now you can test with: apt-get install liblz4-dev

bcharrow commented:
That last couple of actions completes the TODOs.

One question: where are the streaming read functions actually called?  Grepping through ros_comm, it looks like Stream::startRead() is only called in ChunkedFile::setReadMode(), and that that's only called in ChunkedFile:seek() which always uses the uncompressed option.  Are startRead() for BZ2Stream and LZ4Stream actually unused or am I missing something?

bcharrow commented:
It looks like the build failed due to a git checkout error: "FATAL: Could not checkout null with start point dc400a09d95f4af3e25d961cd209e60ece1d0147"?  Not sure what that hash refers to.

bcharrow commented:
Never mind, that's the merge commit.  

@dirk-thomas Do you know how I can fix the build?

dirk-thomas commented:
@ros-pull-request-builder please retest this

ros-pull-request-builder commented:
Test passed.
Refer to this link for build results: http://jenkins.ros.org/job/_pull_request-indigo-ros_comm/5/

dirk-thomas commented on tools/rosbag_storage/src/lz4_stream.cpp:
```diff
@@ -0,0 +1,204 @@
+/*********************************************************************
+* Software License Agreement (BSD License)
+*
+*  Copyright (c) 2008, Willow Garage, Inc.
```
That should be your copyright?

dirk-thomas commented on utilities/roslz4/CMakeLists.txt:
```diff
@@ -0,0 +1,44 @@
+  LIBRARY DESTINATION ${CMAKE_INSTALL_PREFIX}/${PYTHON_INSTALL_DIR}/roslz4)
+
+install(DIRECTORY include/${PROJECT_NAME}/
+  DESTINATION ${CATKIN_GLOBAL_INCLUDE_DESTINATION}
```
Without having tries the install this rule looks problematic. The trailing slash of `include/${PROJECT_NAME}/` only installs the content of that folder but the destination is the global include folder? I think it should either remove the slash to install the folder itself or change the destination to the package specific include destination.

dirk-thomas commented on utilities/roslz4/CMakeLists.txt:
```diff
@@ -0,0 +1,44 @@
+target_link_libraries(roslz4 ${lz4_LIBRARY} ${catkin_LIBRARIES})
+
+# Python bindings
+set(Python_ADDITIONAL_VERSIONS 2.7)
```
Do not use a hard coded Python version here.

Instead use the following two lines:

```
set(Python_ADDITIONAL_VERSIONS "${PYTHON_VERSION_MAJOR}.${PYTHON_VERSION_MINOR}")
set(PythonLibs_FIND_VERSION "${PYTHON_VERSION_MAJOR}.${PYTHON_VERSION_MINOR}")
```

bcharrow replied:
The C module currently uses python 2.  Should I follow tf2_py's lead and use these lines?

```
find_package(PythonLibs 2 REQUIRED)
include_directories(${PYTHON_INCLUDE_PATH})
```

dirk-thomas replied:
It should definitely also compile again Python 3.x. We made huge efforts in the last half year to get the code base to support both and it would be said to introduce new stuff which makes it not build again.

dirk-thomas commented on utilities/roslz4/CMakeLists.txt:
```diff
@@ -0,0 +1,44 @@
+find_package(catkin COMPONENTS)
+
+find_path(lz4_INCLUDE_DIR NAMES lz4.h)
+find_library(lz4_LIBRARY NAMES lz4)
```
If those two resources are in non standard locations you will also need to export those via catkin_package().

dirk-thomas commented:
Beside my tiny comments the PR looks great and I am looking forward to merge it. Thank you, Ben.

bcharrow commented:
I think that should do it.

Thanks for the helpful feedback, Dirk.

dirk-thomas commented:
Please rebase the PR to make it mergable. Also consider squashing the commits into one or a few commits.

bcharrow committed:
Add roslz4 package

Contains C++ and python libraries for v1.4 of the lz4 streaming format.
Both libraries use the system version of lz4 and use default values for
the stream parameters (e.g., includes a stream checksum, but not block
checksums).  See:
http://fastcompression.blogspot.fr/2013/04/lz4-streaming-format-final.html

The C++ library provide a stream style API for
compressing/decompressing data, similar to the bzip library.  They also
contain convenience buffer to buffer compression/decompression.

The python library provides an API similar to the bz2 module and works
with python 2 and 3.

bcharrow committed:
Add lz4 compression to rosbag (python and C++)

bcharrow committed:
Add lz4 compression option in rosbag command line tools

rosbag compress and record now take '--lz4' as an option.  Defaults are
the same; rosbag compress uses bz2 and rosbag record uses nothing.

"rosbag compress" and "rosbag record", implemented by rosbag_main.py, use
the last flag to determine the compression option (e.g., '--bz2 --lz4
--bz2' means bz2 will be used).  This choice is consistent with other
python commands (e.g., "rosbag compress --output-dir=/foo
--output-dir=/tmp file.bag" will put the compressed bag in /tmp).

In contrast, the C++ record executable will raise an error if it is
passed multiple compression types.  This is consistent with boost's
program_options which by default forbids multiple occurrences of the
same flag.

bcharrow commented:
Rebased and condensed into 3 commits.

dirk-thomas commented:
Thank you very much for this significant contributions and the effort to iterate on it.
